docs: extract sections in how to write integration tests to their own howto guide (#2662)
canonical/operator: docs/howto/manage-resources.md

Reviewer: I don't think this comment is needed. We now have a how-to guide "How to publish your charm on Charmhub" and Charmcraft has guidance about publishing resources.

Reviewer: I'd remove this sentence. It duplicates the sentence after the example. I'd replace it by something like:

> A charm can require `file` or `oci-image` resources, defined in `charmcraft.yaml`. When you publish your charm, you'll upload the resources to Charmhub. Then when a user deploys your charm from Charmhub, the resources will be available to your charm code.

Author: I agree this sentence is duplicated. I will remove it.

I like your suggestion, but I might modify it a bit. From what I understand, publishing resources to Charmhub is an explicit step after you publish a charm, and we don't want users to misunderstand that they are published together automatically.

> A charm can require `file` or `oci-image` resources, defined in `charmcraft.yaml`. After you publish your charm, you must explicitly upload the resources to Charmhub. Then when a user deploys your charm from Charmhub, the resources will be available to your charm code.

If `must explicitly` is too strong, perhaps I can use `need to` instead :thinking: .

Reviewer: How about something like:

> As one of the charm publishing steps, you'll upload the resources to Charmhub.

*Edit:* Or maybe it reads better the other way round:

> You'll upload the resources to Charmhub as one of the charm publishing steps.

Author: > You'll upload the resources to Charmhub as one of the charm publishing steps.

Your suggestion is much better. I will go with it. Thank you.

    path = mgr.charm.model.resources.fetch('foo')
    assert path == pathlib.Path('/path/to/resource.tar')
```

(manage-resources-integration-tests)=
### Write integration tests

> See first: {ref}`write-integration-tests-for-a-charm`

During development and testing, it's useful to specify resource locations when deploying the charm.

> The conventional place to specify resource locations for testing is the `upstream-source` field in `charmcraft.yaml`'s `resources` section:

Reviewer: Reading this again, it sounds like we're about to show `charmcraft.yaml`. If you think there's potential for confusion, we could add an extra sentence:

```suggestion
The conventional place to specify resource locations for testing is the `upstream-source` field in `charmcraft.yaml`'s `resources` section. To use `upstream-source` in an integration test:
```

I'm not convinced this addition is needed - up to you.

Author: Interesting. I didn't recall having this sentence in the latest commit. 

I checked and the latest state of this branch didn't have it, so it's okay. I don't think we need it too.

    path = mgr.charm.model.resources.fetch('foo')
    assert path == pathlib.Path('/path/to/resource.tar')
```

(manage-resources-integration-tests)=
### Write integration tests

> See first: {ref}`write-integration-tests-for-a-charm`

A charm can require `file` or `oci-image` resources to work, which have revision numbers on Charmhub. OCI images can be referenced directly, while file resources are typically built during packing.

> In `charmcraft.yaml`'s `resources` section, the `upstream-source` field is, by convention, a usable default image or file that can be used in testing:

Reviewer: ```suggestion
During development and testing, it's useful to specify resource locations when deploying the charm.

The conventional place to specify resource locations for testing is the `upstream-source` field in `charmcraft.yaml`'s `resources` section:
```

```python
import pathlib

import jubilant
import pytest
import yaml

METADATA = yaml.safe_load(pathlib.Path('./charmcraft.yaml').read_text())

@pytest.mark.juju_setup
def test_deploy(charm: pathlib.Path, juju: jubilant.Juju):
    resources = {
        name: res['upstream-source']
        for name, res in METADATA['resources'].items()
    }
    juju.deploy(charm, resources=resources)
    juju.wait(jubilant.all_active)
```

See also: [](jubilant.Juju.deploy)

> Examples of deploying with resources using Jubilant: [`valkey-operator`](https://github.com/canonical/valkey-operator/blob/9/edge/tests/integration/test_charm.py), [`kafka-k8s-operator`](https://github.com/canonical/kafka-k8s-operator/blob/main/tests/integration/test_balancer.py)

Reviewer: I think the context should be clear enough

```suggestion
Examples: [`valkey-operator`](https://github.com/canonical/valkey-operator/blob/9/edge/tests/integration/test_charm.py), [`kafka-k8s-operator`](https://github.com/canonical/kafka-k8s-operator/blob/main/tests/integration/test_balancer.py)
```
